src/store.rs

```diff
@@ -8,130 +9,159 @@ use zkevm_opcode_defs::{
+    pub is_service: bool,
+    pub address: H160,
+    pub shard_id: u8,
+    pub tx_number: u16,
+}
+
+pub trait InitialStorage: Debug {
     fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError>;
-    fn storage_write(&mut self, key: StorageKey, value: U256) -> Result<(), StorageError>;
-    fn storage_drop(&mut self, key: StorageKey) -> Result<(), StorageError>;
-    fn get_state_storage(&self) -> &HashMap<StorageKey, U256>;
-    fn get_all_keys(&self) -> Vec<StorageKey>;
-    fn fake_clone(&self) -> InMemory;
-    fn rollback(&mut self, previous: &dyn Storage) {
-        let keys = previous.get_all_keys();
-        for key in keys {
-            let value = previous.storage_read(key).unwrap();
-            if let Some(value) = value {
-                self.storage_write(key, value).unwrap();
-            }
-        }
-        let current_keys = self.get_all_keys();
-        for key in current_keys {
-            let res = previous.storage_read(key);
-            if let Ok(None) = res {
-                self.storage_drop(key).unwrap();
-            };
-        }
-    }
 }
 
-/// Error type for storage operations.
-#[derive(Error, Debug)]
-pub enum StorageError {
-    #[error("Key not present in storage")]
-    KeyNotPresent,
-    #[error("Error writing to storage")]
-    WriteError,
-    #[error("Error reading from storage")]
-    ReadError,
+#[derive(Debug, Clone)]
+pub struct InitialStorageMemory {
+    pub initial_storage: HashMap<StorageKey, U256>,
 }
 
-/// In-memory storage implementation.
-#[derive(Debug, Clone, Default)]
-pub struct InMemory {
-    contract_storage: HashMap<U256, Vec<U256>>,
-    pub state_storage: HashMap<StorageKey, U256>,
+impl InitialStorage for InitialStorageMemory {
+    fn storage_read(&self, key: StorageKey) -> Result<Option<U256>, StorageError> {
+        Ok(self.initial_storage.get(&key).copied())
+    }
 }
 
-#[derive(Clone, Copy, PartialEq, Eq, Hash, Debug)]
-pub struct StorageKey {
-    pub address: H160,
-    pub key: U256,
+pub trait ContractStorage: Debug {
+    fn decommit(&self, hash: U256) -> Result<Option<Vec<U256>>, StorageError>;
+}
+#[derive(Debug)]
+pub struct ContractStorageMemory {
+    pub contract_storage: HashMap<U256, Vec<U256>>,
 }
 
-impl StorageKey {
-    pub fn new(address: H160, key: U256) -> Self {
-        Self { address, key }
+impl ContractStorage for ContractStorageMemory {
+    fn decommit(&self, hash: U256) -> Result<Option<Vec<U256>>, StorageError> {
+        Ok(self.contract_storage.get(&hash).cloned())
     }
 }
 
-impl InMemory {
-    pub fn new_empty() -> Self {
-        let state_storage = HashMap::new();
-        let contract_storage = HashMap::new();
-        InMemory {
-            state_storage,
-            contract_storage,
-        }
-    }
+#[derive(Debug)]
+pub struct StateStorage {
+    pub storage_changes: HashMap<StorageKey, U256>,
+    pub initial_storage: Rc<RefCell<dyn InitialStorage>>,
+    l1_to_l2_logs: Vec<L2ToL1Log>,
```

Comment:
The field is` l1_to_l2_logs` but the type is L2ToL1Log, so which one is it then?